src/main/scala/com/microsoft/hyperspace/index/IndexLogEntryTags.scala

```diff
@@ -53,4 +53,8 @@ object IndexLogEntryTags {
   // for Hybrid Scan.
   val INMEMORYFILEINDEX_HYBRID_SCAN_APPENDED: IndexLogEntryTag[InMemoryFileIndex] =
   IndexLogEntryTag[InMemoryFileIndex]("inMemoryFileIndexHybridScanAppended")
+
+  // WHYNOT_REASON stores reason string for disqualification.
+  val WHYNOT_REASON: IndexLogEntryTag[String] =
+    IndexLogEntryTag[String]("whyNotReason")
```

Comment:
Maybe `IndexLogEntryTag[Seq[String]]` instead of `IndexLogEntryTag[String]`?

Reply:
As a side note, not specific to this PR - I don't like the IndexLogEntryTag approach because it bloats IndexLogEntry, makes it mutable, and IndexLogEntry being cached can unnecessarily keep information from earlier query executions which might not be needed again. And maybe I'm wrong, but how does it work when there are multiple concurrent query executions and they reference the same index log entry?

Instead of IndexLogEntryTag, having an execution context object containing such information specific to each execution would be better. It makes IndexLogEntry immutable and allows something that cannot be done now - like caching common information which doesn't depend on IndexLogEntry, such as source data signature.

Reply:
It would be ok since it's only tagged when whyNot flag is enabled.
And the API will be usually called from notebooks, so that might be ok if we clear the tag properly. + hyperspace object is currently ThreadLocal so it won't affect another concurrent query.

In the previous structure - 2 different rules - it's hard to share the context between rules as they are called from Spark optimizer, so we introduced IndexLogEntryTag to cache the execution context.
(fyi Spark3 support Plan https://github.com/microsoft/hyperspace/issues/185)

One good thing of IndexLogEntryTag is it can be shared between different queries having the same source plan key. 

We could introduce another execution context as you said, after this refactoring, but let's keep the index log entry tag approach for faster refactoring.